src/leaflike/db.clj

```diff
@@ -23,31 +27,43 @@
   (let [body (-> request :body)
         bkm (add-created-at body)]
     (if (is-valid-bookmark? body)
-      (jdbc/insert! (db-spec) :bookmarks bkm)
+      (jdbc/insert! (db-spec) (-> (helpers/insert-into :bookmarks)
+                                  (helpers/values [body])
+                                  sql/format))
       ; else
       {:error "Invalid Data"})))
 
 (defn list-all
   []
-  (jdbc/query (db-spec) ["select * from bookmarks"]))
-
+  (jdbc/query (db-spec) (-> (helpers/select :*)
+                            (helpers/from :bookmarks)
+                            sql/format)))
 
-(defn list-by-id
+(defn list-by-params
   [params]
-  (jdbc/query (db-spec) ["select * from bookmarks
-                          where id = ?"
-                         (Integer/parseInt (:id params))]))
+  (let [id (Integer/parseInt (:id params))
+        title (:title params)]
+    (jdbc/query (db-spec) (-> (helpers/select :*)
+                              (helpers/from :bookmarks)
+                              (helpers/merge-where (when (> id 0)
+                                             [:= :id id]))
+                              (helpers/merge-where (when-not (nil? title)
+                                                     ["like" :title (str \% title \%)]))
+                              sql/format))))
 
 (defn list-bookmark
   [request]
   (let [params (clojure.walk/keywordize-keys (-> request :query-params))]
     (if (empty? params)
```

Comment:
Maybe use `cond` in place of nested `if`s

Reply:
They're interdependent.

Reply:
>**Cond** takes a set of test/expr pairs. It evaluates each test one at a time.  If a test returns logical true, cond evaluates and returns the value of the corresponding expr and doesn't evaluate any of the other tests or exprs.


Try this
> (cond
>     (empty? params)                 (list-all)
>     (is-valid-params? params) (list-by-params params)
>     :else                                      {:error "Invalid Params})